File: sdk/swift/Sources/EasyNetDaemonSDK/SDKError.swift
import Foundation

public enum SDKErrorCode: String, Sendable {
    case invalidArgument = "INVALID_ARGUMENT"
    case invalidHandle = "INVALID_HANDLE"
    case daemonOffline = "DAEMON_OFFLINE"
    case permissionDenied = "PERMISSION_DENIED"
    case admissionDenied = "ADMISSION_DENIED"
    case httpAuthDenied = "HTTP_AUTH_DENIED"
    case signatureDenied = "SIGNATURE_DENIED"
    case policyDenied = "POLICY_DENIED"
    case authorityDenied = "AUTHORITY_DENIED"
    case abilityNotFound = "ABILITY_NOT_FOUND"
    case routeUnavailable = "ROUTE_UNAVAILABLE"
    case executionFailed = "EXECUTION_FAILED"
    case timeout = "TIMEOUT"
    case cancelled = "CANCELLED"
    case invalidInvocation = "INVALID_INVOCATION"
    case protocolMismatch = "PROTOCOL_MISMATCH"
    case versionIncompatible = "VERSION_INCOMPATIBLE"
    case controlOnly = "CONTROL_ONLY"
    case transport = "TRANSPORT"
    case protocolFailure = "PROTOCOL"
    case notImplemented = "NOT_IMPLEMENTED"
    case generic = "GENERIC"
}

public enum SDKErrorClass: String, Sendable {
    case validation
    case handle
    case availability
    case permission
    case admission
    case routing
    case timeout
    case cancellation
    case protocolFailure = "protocol"
    case version
    case control
    case unsupported
    case generic
}

public enum RetryHint: String, Sendable {
    case never
    case safe
    case afterBackoff = "after_backoff"
    case unknown
}

public struct SDKError: Error, Sendable, CustomStringConvertible {
    public let code: SDKErrorCode
    public let stage: String
    public let retryHint: RetryHint
    public let retryable: Bool
    public let message: String
    public let source: String
    public let invocationID: String
    public let receiptURA: String
    public let details: [String: String]

    public init(
        code: SDKErrorCode,
        stage: String,
        retryHint: RetryHint = .never,
        retryable: Bool = false,
        message: String,
        source: String = "",
        invocationID: String = "",
        receiptURA: String = "",
        details: [String: String] = [:]
    ) {
        self.code = code
        self.stage = SDKError.required(stage, "stage")
        self.retryHint = retryHint
        self.retryable = retryable
        self.message = SDKError.required(message, "message")
        self.source = source
        self.invocationID = invocationID
        self.receiptURA = receiptURA
        self.details = details
    }

    public static func validation(_ stage: String, _ message: String) -> SDKError {
        SDKError(code: .invalidArgument, stage: stage, message: message)
    }

    public static func closed(_ stage: String) -> SDKError {
        SDKError(code: .invalidHandle, stage: stage, message: "\(stage) is closed")
    }

    public var errorClass: SDKErrorClass {
        switch code {
        case .invalidArgument, .invalidInvocation:
            return .validation
        case .invalidHandle:
            return .handle
        case .daemonOffline, .routeUnavailable, .transport:
            return .availability
        case .permissionDenied, .httpAuthDenied:
            return .permission
        case .admissionDenied, .signatureDenied, .policyDenied, .authorityDenied, .executionFailed:
            return .admission
        case .abilityNotFound:
            return .routing
        case .timeout:
            return .timeout
        case .cancelled:
            return .cancellation
        case .protocolFailure, .protocolMismatch:
            return .protocolFailure
        case .versionIncompatible:
            return .version
        case .controlOnly:
            return .control
        case .notImplemented:
            return .unsupported
        case .generic:
            return .generic
        }
    }

    public var description: String {
        "\(code.rawValue): \(message)"
    }

    static func required(_ value: String?, _ field: String) -> String {
        guard let value, !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return field
        }
        return value
    }
}

File: sdk/java/src/main/java/run/easynet/daemon/SDKError.java
package run.easynet.daemon;

import java.util.Map;
import java.util.Objects;

public final class SDKError extends RuntimeException {
  private static final long serialVersionUID = 1L;

  private final ErrorCode code;
  private final String stage;
  private final RetryHint retryHint;
  private final boolean retryable;
  private final String source;
  private final String invocationId;
  private final String receiptURA;
  private final transient Map<String, Object> details;

  public SDKError(
      ErrorCode code,
      String stage,
      RetryHint retryHint,
      boolean retryable,
      String message,
      String source,
      String invocationId,
      String receiptURA,
      Map<String, Object> details,
      Throwable cause) {
    super(required(message, "message"), cause);
    this.code = Objects.requireNonNull(code, "code");
    this.stage = required(stage, "stage");
    this.retryHint = Objects.requireNonNullElse(retryHint, RetryHint.NEVER);
    this.retryable = retryable;
    this.source = source == null ? "" : source;
    this.invocationId = invocationId == null ? "" : invocationId;
    this.receiptURA = receiptURA == null ? "" : receiptURA;
    this.details = details == null ? Map.of() : Map.copyOf(details);
  }

  public static SDKError validation(String stage, String message) {
    return new SDKError(
        ErrorCode.INVALID_ARGUMENT,
        stage,
        RetryHint.NEVER,
        false,
        message,
        "",
        "",
        "",
        Map.of(),
        null);
  }

  public static SDKError closed(String stage) {
    return new SDKError(
        ErrorCode.INVALID_HANDLE,
        stage,
        RetryHint.NEVER,
        false,
        stage + " is closed",
        "",
        "",
        "",
        Map.of(),
        null);
  }

  public ErrorCode code() {
    return code;
  }

  public String stage() {
    return stage;
  }

  public RetryHint retryHint() {
    return retryHint;
  }

  public boolean retryable() {
    return retryable;
  }

  public String source() {
    return source;
  }

  public String invocationId() {
    return invocationId;
  }

  public String receiptURA() {
    return receiptURA;
  }

  public Map<String, Object> details() {
    return details;
  }

  public ErrorClass errorClass() {
    return switch (code) {
      case INVALID_ARGUMENT, INVALID_INVOCATION -> ErrorClass.VALIDATION;
      case INVALID_HANDLE, NOT_INITIALIZED -> ErrorClass.HANDLE;
      case ROUTE_UNAVAILABLE, TRANSPORT -> ErrorClass.AVAILABILITY;
      case PERMISSION_DENIED, HTTP_AUTH_DENIED -> ErrorClass.PERMISSION;
      case ADMISSION_DENIED, SIGNATURE_DENIED, POLICY_DENIED, AUTHORITY_DENIED,
              EXECUTION_FAILED ->
          ErrorClass.ADMISSION;
      case ABILITY_NOT_FOUND -> ErrorClass.ROUTING;
      case TIMEOUT -> ErrorClass.TIMEOUT;
      case CANCELLED -> ErrorClass.CANCELLATION;
      case PROTOCOL, PROTOCOL_MISMATCH -> ErrorClass.PROTOCOL;
      case VERSION_INCOMPATIBLE -> ErrorClass.VERSION;
      case CONTROL_ONLY -> ErrorClass.CONTROL;
      case NOT_IMPLEMENTED -> ErrorClass.UNSUPPORTED;
      case GENERIC -> ErrorClass.GENERIC;
    };
  }

  private static String required(String value, String field) {
    if (value == null || value.isBlank()) {
      throw new IllegalArgumentException(field + " is required");
    }
    return value;
  }
}
